src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs

```diff
@@ -479,7 +480,15 @@ protected override bool DoTryComplete(SocketAsyncContext context)
                 }
                 else
                 {
-                    return SocketPal.TryCompleteReceiveFrom(context._socket, Buffer.Span, null, Flags, SocketAddress, ref SocketAddressLen, out BytesTransferred, out ReceivedFlags, out ErrorCode);
+                    if (!SetReceivedFlags && SocketAddress == null)
```

Comment:
In what situation is SetReceivedFlags false but SocketAddress != null?  I'm wondering why we need both of these checks and/or if SetReceivedFlags could be initialized appropriately based on SocketAddress.

Reply:
> In what situation is SetReceivedFlags false but SocketAddress != null?

That doesn't happen.

> I'm wondering why we need both of these checks and/or if SetReceivedFlags could be initialized appropriately based on SocketAddress.

I think we need both. Consider UDP socket, when calling `Socket.ReceiveAsync(saea)`, `SetReceivedFlags = true, SocketAddress == null`. And when calling `Socket.ReceiveFromAsync(saea)`, `SetReceivedFlags = true, SocketAddress != null`.

Reply:
But why wouldn't you just set:
SetReceivedFlags = SocketAddress != null;
in the latter case?

Reply:
I've removed `SocketAddress` from the check, and added an Assert instead.

> But why wouldn't you just set:
SetReceivedFlags = SocketAddress != null;
in the latter case?

I don't understand the question/suggestion. Where would this initializer be? Currently initialization is hard-coded to true/false.

Reply:
> Where would this initializer be? Currently initialization is hard-coded to true/false.

Where you currently have it hardcoded to true.  If the only purpose of this was to control which of these paths were taken, and if everywhere it was checked we checked both conditions, then you could just use the one flag to represent both conditions by setting it appropriately based on both.

> I've removed SocketAddress from the check, and added an Assert instead.

If the check wasn't actually necessary and an assert suffices, even better.

Reply:
> Where you currently have it hardcoded to true.

We want it to be `true` also `Socket.ReceiveAsync(saea)` gets called for UDP. Then `SocketAddress` is `null`. (`SocketAsyncContext.ReceiveFromAsync` gets called for `Socket.ReceiveAsync`.)

Reply:
Either I'm not understanding or we're talking past each other.  But it doesn't matter anymore as you removed the extra check.